controllers/jujumachine_controller.go

```diff
@@ -152,11 +195,8 @@ func (r *JujuMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		// registering our finalizer.
 		if !controllerutil.ContainsFinalizer(jujuMachine, infrastructurev1beta1.JujuMachineFinalizer) {
 			controllerutil.AddFinalizer(jujuMachine, infrastructurev1beta1.JujuMachineFinalizer)
-			if err := r.Update(ctx, jujuMachine); err != nil {
-				return ctrl.Result{}, err
-			}
 			log.Info("added finalizer")
-			return ctrl.Result{}, nil
```

Comment:
I think we still need to return immediately after adding the finalizer. See for example the [Cluster controller](https://github.com/kubernetes-sigs/cluster-api/blob/4f60841913fa99912771c3906ee81dd13d4e9633/internal/controllers/cluster/cluster_controller.go#L144-L148). The comment there suggests that there's a race condition to avoid. I think it goes like this:

1. Controller "adds finalizer" but does not update
2. Controller proceeds further into reconcile, creating new resources and such (e.g. an actual machine in Juju)
4. User deletes JujuMachine, which deletes immediately because it has no finalizers
5. Controller finishes reconcile, tries to update the JujuMachine, but fails because it no longer exists
6. Resources created in step 2 never get deleted

So yeah, I think we need to return here, let the resource get updated, and then only proceed further once the finalizer is surely in place.

Reply:
yeah nice catch. Makes sense that we dont wanna go through everything until things are safely deletable. Ive added the return back in